### PR TITLE
Move cleanup record test to upgrade schedule

### PR DESCRIPTION
**Pull request by agedemenli**
DESCRIPTION: Move cleanup record test to upgrade schedule

### CODE REVIEW

File: src/test/regress/sql/multi_extension.sql [hunk 1]
```diff
@@ -530,32 +530,14 @@ SELECT * FROM multi_extension.print_extension_changes();
...
-UPDATE pg_dist_placement SET shardstate = 4 WHERE shardid % 2 = 1;
-
 ALTER EXTENSION citus UPDATE TO '11.1-1';
 
--- show placements and cleanup records
-SELECT * FROM pg_dist_placement ORDER BY shardid;
-SELECT * FROM pg_dist_cleanup;
-
 -- Should be empty result since upgrade+downgrade should be a no-op
 SELECT * FROM multi_extension.print_extension_changes();
```
Comment (thanodnl) [inline src/test/regress/sql/multi_extension.sql:536]: small nit; at lines 532 and 534 we have empty lines. If we look at the block above L517-521 we see a concise pattern of upgrade-downgrade-test. Would be good to keep the style of the test the same, eg. remove the empty lines in this block.

File: src/test/regress/sql/multi_extension.sql [hunk 2]
```diff
@@ -530,32 +530,14 @@ SELECT * FROM multi_extension.print_extension_changes();
...
 -- Snapshot of state at 11.2-1
 ALTER EXTENSION citus UPDATE TO '11.2-1';
 
--- verify that the placements are deleted and cleanup records are created
-SELECT * FROM pg_dist_placement ORDER BY shardid;
-SELECT * FROM pg_dist_cleanup;
-
 ALTER EXTENSION citus_columnar UPDATE TO '11.2-1';
 
 -- Make sure that we defined dependencies from all rel objects (tables,
```
Comment (thanodnl) [inline src/test/regress/sql/multi_extension.sql:543]: small note on this, I see we also perform some columnar checks here with regards to an upgrade. I guess these should be moved to actual upgrade tests as well. Possibly in a separate PR, and possibly by the author of this block (hopes that is not me ;) )

File: src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql [hunk 3]
```diff
@@ -0,0 +1,20 @@
+\set upgrade_test_old_citus_version `echo "$CITUS_OLD_VERSION"`
+SELECT substring(:'upgrade_test_old_citus_version', 'v(\d+)\.\d+\.\d+')::int < 10
+AS upgrade_test_old_citus_version_lt_10_0;
+\gset
+\if :upgrade_test_old_citus_version_lt_10_0
+\else
+\q
```
Comment (thanodnl) [inline src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql:7]: could we leave a comment here on why we are _not_ testing this when the version before an upgrade is citus 10 or greater.

Also, since we patched something in Citus 11.2 for this, are we testing for the correct Citus version here?

Lastly, given we test a version and then do nothing in the if-block, and only have 'code' in the else block, would it make sense to inverse the check we are performing?

File: src/test/regress/sql/upgrade_basic_after.sql [hunk 4]
```diff
@@ -8,8 +8,18 @@ SELECT nextval('pg_dist_placement_placementid_seq') = MAX(placementid)+1 FROM pg
...
+    THEN true
+    ELSE nextval('pg_dist_operationid_seq') = MAX(operation_id)+1
+    END AS check_operationid
+    FROM pg_dist_cleanup;
+SELECT
+    CASE WHEN MAX(record_id) IS NULL
+    THEN true
+    ELSE nextval('pg_dist_cleanup_recordid_seq') = MAX(record_id)+1
+    END AS check_recordid
+    FROM pg_dist_cleanup;
```
Comment (thanodnl) [inline src/test/regress/sql/upgrade_basic_after.sql:22]: this change seems unrelated on first sight, why do we need to rewrite these tests now?

Instead, I see some references to pg_dist_cleanup in the original PR, are we uncovering/hiding an upgrade bug here? Shouldn't the sequences retain the correct value by accompanying logic in `citus_(prepare|finish)_pg_upgrade`?

**Review by thanodnl — COMMENTED** [review]
There are some questions surrounding the cleanup sequence and operation sequence which need to be addressed.

Besides there are some small nits around style of the code, which make the code easier to read and understand.

File: src/test/regress/sql/upgrade_basic_after.sql (hunk 4 above)
Comment (agedemenli) [reply src/test/regress/sql/upgrade_basic_after.sql:22]: this test file is called in two schedules: citus upgrade and the pg upgrade one.
We create cleanup record with the test in this pr in citus upgrade schedule, not in pg upgrade schedule.
In pg upgrade schedule, because no records are created, the checks for these sequences return null, but in the citus upgrade schedule, they return true. That's why I changed these sequence checks so that they are expected to be either null, or max+1.

File: src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql (hunk 3 above)
Comment (agedemenli) [reply src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql:7]: afaik we have two versions to use in the upgrade tests.
v9.5.0 and v10.1.0
Current patch runs the before test for v9.5 and then runs the after test for the latest version. Since the cleanup record changes were introduced after v9.5, we are able to test it by comparing with v9.5. We could've done it for v10.1.0 as well and probably nothing would change. 

I'm not sure I've entirely understood your comment.
Can you explain which patch have you mentioned?

File: src/test/regress/sql/multi_extension.sql (hunk 2 above)
Comment (agedemenli) [reply src/test/regress/sql/multi_extension.sql:543]: @onurctirtir  does it make sense to you to move columnar check to the upgrade schedule?

File: src/test/regress/sql/upgrade_basic_after.sql (hunk 4 above)
Comment (thanodnl) [reply src/test/regress/sql/upgrade_basic_after.sql:22]: > In pg upgrade schedule, because no records are created, the checks for these sequences return null

On deeper inspection I do see that, even though the tests look very much like the tests above, the output used to be `NULL` for these two lines, compared to the expected true value.

This is probably ok, but the deviation from the tests above warrants a comment describing the situation where the `MAX(operation_id) IS NULL`. Given the precedent above where sequences are generally tested the simple way, like before this PR -albeit that test was wrong for these sequences and should not have been added given the wrong output.

However, I don't think it is easy or possible, but having `pg_dist_cleanup` populated during postgres upgrades would be even better. Although I don't think that would be a valid scenario, but that would be great to get affirmed. (do understand that I don't really understand the initial change you were making - maybe it's better to involve the people that reviewed https://github.com/citusdata/citus/pull/6494)

File: src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql (hunk 3 above)
Comment (thanodnl) [reply src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql:7]: > afaik we have two versions to use in the upgrade tests.
> v9.5.0 and v10.1.0

This is currently the case, but once in a blue moon we might revisit the versions we test upgrades against. Don't assume these will be the only version we ever test against. Our version checks need to make sense, not only discriminate between the two current versions we test against.

> Can you explain which patch have you mentioned?

This PR is changing tests we have added via https://github.com/citusdata/citus/pull/6494 which seems to have been landed in 11.2, therefor it seems strange to me that we don't have to test this if Citus originates from v10.x.y. I would assume an upgrade from 10 to 11.2 would also require this `before` and `after` steps to run. Given we specifically quit the test here if the before version is v10 or above it would be great to add a comment on _why_ setting up is either not possible, or will not trigger the desired conditions we are testing.

Comment (JelteF) [reply src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql:7]: The check in this file should be changed to a before Citus 11.2 check instead.

P.S. To clarify how we'll revisit the version upgrades we test: We'll never run any upgrade tests for Citus 9.4 or smaller, because those have no overlapping PG version support with the current Citus version. But in the future we'll want to be able upgrade from later versions when we drop PG support for the last version that supported v10.1.0.

File: src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql [hunk 5]
```diff
@@ -0,0 +1,23 @@
+\set upgrade_test_old_citus_version `echo "$CITUS_OLD_VERSION"`
+SELECT substring(:'upgrade_test_old_citus_version', 'v(\d+)\.\d+\.\d+')::int < 11 OR
+       (substring(:'upgrade_test_old_citus_version', 'v(\d+)\.\d+\.\d+')::int = 11 AND
+        substring(:'upgrade_test_old_citus_version', 'v\d+\.(\d+)\.\d+')::int < 2)
+AS upgrade_test_old_citus_version_lt_11_2;
+\gset
+\if :upgrade_test_old_citus_version_lt_11_2
+\else
+\q
+\endif
```
Comment (thanodnl) [inline src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql:10]: The check still has an empty `\else` block which makes the test hard to reason about. I was looking for something like below:

```suggestion
\set upgrade_test_old_citus_version `echo "$CITUS_OLD_VERSION"`
SELECT substring(:'upgrade_test_old_citus_version', 'v(\d+)\.\d+\.\d+')::int > 11 OR
       (substring(:'upgrade_test_old_citus_version', 'v(\d+)\.\d+\.\d+')::int = 11 AND
        substring(:'upgrade_test_old_citus_version', 'v\d+\.(\d+)\.\d+')::int >= 2)
AS upgrade_test_old_citus_version_gte_11_2;
\gset
\if :upgrade_test_old_citus_version_gte_11_2
\q
\endif
```

File: src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+SET citus.next_shard_id TO 980000;
+CREATE TABLE table_with_orphaned_shards (a int);
+SELECT create_distributed_table('table_with_orphaned_shards', 'a');
+-- show all 32 placements are active
+SELECT COUNT(*) FROM pg_dist_placement WHERE shardstate = 1 AND shardid IN (SELECT shardid FROM pg_dist_shard WHERE logicalrelid='table_with_orphaned_shards'::regclass);
+-- mark one shard as orphaned
+INSERT INTO pg_dist_placement(placementid, shardid, shardstate, shardlength, groupid)
+    SELECT nextval('pg_dist_placement_placementid_seq'::regclass), shardid, 4, shardlength, 3-groupid
+    FROM pg_dist_placement
+    WHERE shardid % 32 = 1 AND shardid IN (SELECT shardid FROM pg_dist_shard WHERE logicalrelid='table_with_orphaned_shards'::regclass)
```
Comment (JelteF) [inline src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql:22]: Nit: maybe simplify and just go for:
```suggestion
    WHERE shardid = 980001
```

**Review by JelteF — APPROVED** [review]

File: src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+\endif
+
+-- create a table with orphaned shards to see if orphaned shards will be dropped
+-- and cleanup records will be created for them
+SET citus.next_shard_id TO 980000;
+CREATE TABLE table_with_orphaned_shards (a int);
+SELECT create_distributed_table('table_with_orphaned_shards', 'a');
+-- show all 32 placements are active
+SELECT COUNT(*) FROM pg_dist_placement WHERE shardstate = 1 AND shardid IN (SELECT shardid FROM pg_dist_shard WHERE logicalrelid='table_with_orphaned_shards'::regclass);
+-- mark one shard as orphaned
```
Comment (JelteF) [inline src/test/regress/sql/upgrade_pg_dist_cleanup_before.sql:18]:
```suggestion
-- create an orphaned placement based on an existing one
```

File: src/test/regress/after_citus_upgrade_coord_schedule [hunk 8]
```diff
@@ -1,6 +1,7 @@
 # this schedule is to be run only on coordinators
 
 test: upgrade_citus_finish_citus_upgrade
+test: upgrade_pg_dist_cleanup_after
```
Comment (JelteF) [inline src/test/regress/after_citus_upgrade_coord_schedule:4]: nit: nice to move this to the same as the before test
